python/dcr_template.py

```diff
@@ -1452,3 +1627,71 @@ def wrap_warpExposure(exposure, wcs, BBox, warpingControl=None):
     warpVariance = warpExp.getMaskedImage().getVariance().getArray()
     exposure.getMaskedImage().getVariance().getArray()[:, :] = warpVariance
     exposure.setWcs(wcs)
+
+
+def solve_model(kernel_size, img_vals, n_step, kernel_dcr, kernel_ref=None, kernel_restore=None):
+    """!Wrapper to call a fitter using a given covariance matrix, image values, and any regularization.
+
+    Parameters
+    ----------
+    kernel_size : int
+        Size of the kernel to use for calculating the covariance matrix, in pixels.
+    img_vals : np.ndarray
+        Image data values for the pixels being used for the calculation, as a 1D vector.
+    n_step : int, optional
+        Number of sub-filter wavelength planes to model.
+    kernel_dcr : np.ndarray
+        The covariance matrix describing the effect of DCR
+    kernel_ref : np.ndarray, optional
+        The covariance matrix for the reference image
+    kernel_restore : np.ndarray, optional
+        The covariance matrix for the final restored image
+
+    Returns
+    -------
+    np.ndarray
+        Array of the solution values.
+    """
+    x_size = kernel_size
+    y_size = kernel_size
+    if (kernel_restore is None) or (kernel_ref is None):
+        vals_use = img_vals
+        kernel_use = kernel_dcr
+    else:
+        vals_use = kernel_restore.dot(img_vals)
+        kernel_use = kernel_ref.dot(kernel_dcr)
+
+    model_solution = scipy.optimize.nnls(kernel_use, vals_use)
+    model_vals = model_solution[0]
+    n_pix = x_size*y_size
+    for f in range(n_step):
+        yield np.reshape(model_vals[f*n_pix: (f + 1)*n_pix], (y_size, x_size))
+
+
+def calculate_rotation_angle(exposure):
+    """Calculate the sky rotation angle of an exposure.
+
+    Parameters
+    ----------
+    exposure : lsst.afw.image.ExposureD
+        An LSST exposure object.
+
+    Returns
+    -------
+    lsst.afw.geom.Angle
+        The rotation of the image axis, East from North.
+    """
+    visitInfo = exposure.getInfo().getVisitInfo()
+
+    az = visitInfo.getBoresightAzAlt().getLongitude()
+    hour_angle = visitInfo.getBoresightHourAngle()
+    if np.isfinite(hour_angle.asRadians()):
+        dec = visitInfo.getBoresightRaDec().getDec()
+        lat = visitInfo.getObservatory().getLatitude()
+        p_angle = parallactic_angle(hour_angle, dec, lat)
+    else:
```

Comment:
I've added a printed warning for now. I'll re-create the simulated data that prompted this branch as DM-9900, and then I can switch to raising an exception as you suggest.